File: backends/java/scenarios/crashproof/CrashProducer.java
// The java half of the crash-recovery proof (roadmap item 322, Slice 2): the
// java analog of the py tier's tests/test_crash_recovery.py "simulated kill -9"
// case and the go tier's crash_proof_test.go, but driven by a REAL JVM process
// that writes a durable WAL and dies.
//
// This is a manual driver, not a JUnit test: it is built and driven by
// tests/test_java_crash_recovery.py (the python half), which sets the
// environment, runs it as a subprocess, then reads the WAL back through the
// tier-agnostic `revl recover`. Run standalone with no REVL_WAL it self-skips.
//
// It boots the witnessed composition (Crasher) on the in-repo cordis4j STUB
// Context — the same seam RunMethodWitnessedH1 uses — whose activation registers
// a witnessed transactional mutation. The recording preamble (emit --record)
// writes that mutation's discharge-descriptor durably to $REVL_WAL and fsyncs it
// (FileChannel.force) BEFORE apply() returns, so the inverse is re-issuable from
// the log alone after the process dies.

import io.cordis4j.core.Context;
import io.cordis4j.core.Disposable;

public final class CrashProducer {
    // Two modes, selected by env (set by the python driver):
    //
    //   - REVL_CRASH_BEFORE_COMMIT=1 — simulate an abrupt process crash:
    //     Runtime.halt the moment the mutation is durable, BEFORE the discharge /
    //     activation-complete markers. The WAL carries the descriptor but no
    //     `discharge` / `activation-complete`: recover rolls back.
    //   - unset — the clean control: dispose (the in-process frame committed at
    //     activation-end, so the witnessed inverse discharges rather than
    //     replaying), then stamp `discharge` + `activation-complete` exactly as
    //     the py driver does: recover skips the committed seq / rolls forward.
    public static void main(String[] args) throws Exception {
        String wal = System.getenv("REVL_WAL");
        if (wal == null || wal.isEmpty()) {
            System.err.println("REVL_WAL not set: this producer runs under the "
                    + "python crash-recovery driver");
            System.exit(2);
        }

        Context root = new Context();
        Disposable activation = new revl.Components.CrasherPlugin().apply(root);
        // The witnessed mutation ran during apply() and its discharge-descriptor
        // is now durable on disk (revlRecordTransactional fsynced it before
        // apply() returned).

        if ("1".equals(System.getenv("REVL_CRASH_BEFORE_COMMIT"))) {
            // Abrupt death mid-session: no discharge record, no terminal marker.
            // halt() (unlike System.exit) skips shutdown hooks and finalizers —
            // the fsynced descriptor is all that survives, exactly what recover
            // must roll back from.
            Runtime.getRuntime().halt(137);
        }

        // Clean control: unload (dispose runs the in-process LIFO teardown; the
        // frame committed at activation-end so the witnessed inverse discharges
        // rather than replaying), then stamp the commit-path proof and the
        // terminal marker, mirroring the py driver's discharge + commit_activation
        // and go's revlRecordDischarge / revlRecordActivationComplete.
        activation.dispose();
        revl.Components.revlRecordDischarge();
        revl.Components.revlRecordActivationComplete();
    }

    private CrashProducer() {}
}

File: backends/rust/scenarios/crashproof/src/bin/crash_producer.rs
//! The rust half of the crash-recovery proof (roadmap item 322, Slice 2): the
//! rust analog of the py tier's tests/test_crash_recovery.py "simulated kill -9"
//! case, but driven by a REAL process that writes a durable WAL and dies.
//!
//! Built and driven by tests/test_rust_crash_recovery.py (the python half),
//! which sets the environment, runs this binary as a subprocess, then reads the
//! WAL back through `revl recover`. Run standalone with no REVL_WAL it self-skips
//! (exit 3, the runtime-unavailable convention — never a green run that recorded
//! nothing).
//!
//! Boots the witnessed composition (Crasher) whose activation registers a
//! witnessed transactional mutation. The recording preamble (emit --record)
//! writes that mutation's discharge-descriptor durably to $REVL_WAL and fsyncs
//! it (`File::sync_all`) BEFORE `wait()` returns, so the inverse is re-issuable
//! from the log alone after this process dies.
//!
//! Two modes, selected by env (set by the python driver):
//!
//!   - REVL_CRASH_BEFORE_COMMIT=1 — simulate an abrupt process crash:
//!     `process::exit(137)` the moment the mutation is durable, BEFORE commit.
//!     The WAL carries the descriptor but no `discharge` / `activation-complete`:
//!     recover rolls back.
//!   - unset — the clean control: dispose (commit -> the witnessed inverse
//!     discharges, never replays), then stamp `discharge` + `activation-complete`
//!     exactly as the py driver does: recover skips the committed seq / rolls
//!     forward.

use std::process::exit;

fn main() {
    if std::env::var("REVL_WAL")
        .ok()
        .filter(|v| !v.is_empty())
        .is_none()
    {
        eprintln!(
            "REVL_WAL not set: this producer runs under the python crash-recovery driver"
        );
        exit(3);
    }

    let root = cordis::Context::new();
    let cfg = serde_json::json!({ "Crasher": {} });
    let f = revl_crashproof::_revl_load(&root, "crasher", &cfg).expect("load Crasher");
    f.wait().expect("Crasher did not reach ACTIVE");
    // The witnessed mutation ran in-process and its discharge-descriptor is now
    // durable on disk (revl_record_transactional fsynced it before returning).

    if std::env::var("REVL_CRASH_BEFORE_COMMIT").ok().as_deref() == Some("1") {
        // Abrupt death mid-session: no discharge record, no terminal marker.
        // The fsynced descriptor is all that survives — exactly what recover
        // must roll back from.
        exit(137);
    }

    // Clean control: unload LIFO (commit flips the frame, so the witnessed
    // inverse discharges rather than replaying), then stamp the commit-path
    // proof and the terminal marker, mirroring the py driver's discharge +
    // commit_activation.
    f.dispose().expect("dispose Crasher");
    revl_crashproof::revl_record_discharge();
    revl_crashproof::revl_record_activation_complete();
}
